src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs

```diff
@@ -236,85 +268,340 @@ private SelectItem GenerateExpandItem(ExpandTermToken tokenIn)
+                return new SelectExpandClause(new Collection<SelectItem>(), true);
             }
+        }
 
-            SelectExpandClause subSelectExpand;
-            if (tokenIn.ExpandOption != null)
+        /// <summary>
+        /// Process a <see cref="SelectTermToken"/> to identify whether it's a Wildcard path.
+        /// </summary>
+        /// <param name="selectToken">the select token to process.</param>
+        /// <param name="newSelectItem">the built select item to out.</param>
+        private bool ProcessWildcardTokenPath(SelectTermToken selectToken, out SelectItem newSelectItem)
+        {
+            newSelectItem = null;
+            if (selectToken == null || selectToken.PathToProperty == null)
             {
-                subSelectExpand = this.GenerateSubExpand(tokenIn);
+                return false;
             }
-            else
+
+            PathSegmentToken pathToken = selectToken.PathToProperty;
+            if (SelectPathSegmentTokenBinder.TryBindAsWildcard(pathToken, this.Model, out newSelectItem))
             {
-                subSelectExpand = BuildDefaultSubExpand();
+                // * or Namespace.*
+                if (pathToken.NextToken != null)
+                {
+                    throw new ODataException(ODataErrorStrings.SelectExpandBinder_InvalidIdentifierAfterWildcard(pathToken.NextToken.Identifier));
+                }
+
+                VerifyNoQueryOptionsNested(selectToken, pathToken.Identifier);
+
+                return true;
             }
 
-            subSelectExpand = this.DecorateExpandWithSelect(subSelectExpand, currentNavProp, tokenIn.SelectOption, this.CreateBindingState(targetNavigationSource, generatedProperties, collapsed));
+            return false;
+        }
+
+        /// <summary>
+        /// Process a <see cref="PathSegmentToken"/> following any type segments if necessary.
+        /// </summary>
+        /// <param name="tokenIn">the path token to process.</param>
+        private List<ODataPathSegment> ProcessSelectTokenPath(PathSegmentToken tokenIn)
+        {
+            Debug.Assert(tokenIn != null, "tokenIn != null");
 
-            LevelsClause levelsOption = ParseLevels(tokenIn.LevelsOption, currentLevelEntityType, currentNavProp);
-            return new ExpandedNavigationSelectItem(pathToNavProp, targetNavigationSource, subSelectExpand, filterOption, orderbyOption, tokenIn.TopOption, tokenIn.SkipOption, tokenIn.CountQueryOption, searchOption, levelsOption, computeOption, applyOption);
+            List<ODataPathSegment> pathSoFar = new List<ODataPathSegment>();
+            IEdmStructuredType currentLevelType = this.edmType;
+
+            // first, walk through all type segments in a row, converting them from tokens into segments.
+            if (tokenIn.IsNamespaceOrContainerQualified() && !UriParserHelper.IsAnnotation(tokenIn.Identifier))
+            {

```

Comment:
Don't know that it would be worth it for this case, but you could avoid allocating the empty list if you just defined the variable on 474, then pathSoFar = new List<ODataPathSegment>(SelectExpandPathBinder....) in 481, and added an "else {pathSoFar=new List<ODataPathSegment>();" at 489.